### PR TITLE
`licenseChoice` streamlined

### LINKED ISSUE #204: JSON and XML schemas differ for `licenses`
As noted in https://github.com/CycloneDX/cyclonedx-python-lib/issues/365, there is a discrepancy between JSON and XML schemas for CycloneDX v1.4 with how `licenses` are defined.

**JSON**
`licenses` is an Array of either `expression` or `license` - i.e. 0 or more `expression` or 0 or more `license` [see here](https://cyclonedx.org/docs/1.4/json/#components_items_licenses_items_license)

**XML**
`licenses` is an complex type defined as either 0 or more `license` OR 0 or 1 `expression` [see here](https://cyclonedx.org/docs/1.4/xml/#type_licenseChoiceType)

### PR DESCRIPTION
**Pull request by jkowalleck**
fixes #204

proposal to streamline `licenses`: make JSON meat the existing XML spec

this does not change the actual SBOM structure in any way.
it only puts a constraint on the document to have exactly one expression in the list, if there is any expression.

### CODE REVIEW

**Review by madpah — APPROVED** [review]

File: schema/bom-1.5.schema.json [hunk 1]
```diff
@@ -802,28 +798,40 @@
...
-      "oneOf":[
+      "type": "array",
+      "title": "License Choice",
+      "description": "EITHER (a list of SPDX and/or named licenses) OR (a list of up to one SPDX License Expression)",
+      "oneOf": [
         {
-          "required": ["license"]
+          "items": {
+            "type": "object",
+            "title": "License(s)",
```
Comment (jkowalleck) [inline schema/bom-1.5.schema.json:808]: i think i should remove this title here.
it is misleading, if not wrong.

Comment (jkowalleck) [reply schema/bom-1.5.schema.json]: removed the title.

File: schema/bom-1.5.schema.json [hunk 2]
```diff
@@ -845,28 +841,37 @@
...
+            "required": ["license"],
+            "additionalProperties": false,
+            "properties": {
+              "license": {"$ref": "#/definitions/license"}
+            }
+          }
         },
         {
-          "required": ["expression"]
+          "additionalItems": false,
```
Comment (jkowalleck) [inline schema/bom-1.5.schema.json:859]: defined items as a list with exactly one item and set `additionalItems` to false

**Review by stevespringett — APPROVED** [review]

**Comment by johan-boule** [comment]
What do we have to do when the tool raises a LicenseExpressionAlongWithOthersException ?

**Comment by jkowalleck** [comment]
It is expected to be either an expression, or a list of licenses. not both.
In case an expression exists, the list of licenses should be ignored for the following reasons:
expression knows the concatenation type (`AND`, `OR`, `WITHOUT`/`EXCEPT`) of the licenses, and is therefore more verbose.

**Comment by johan-boule** [comment]
I don't understand at all what I'm doing but, if I keep only one licence elements in that conanfile.py, this fixes the error.

**Comment by jkowalleck** [comment]
> I don't understand at all what I'm doing but, if I keep only one licence element in that conanfile.py, this fixes the error.

This is the repository of CycloneDX specification - the internation standard.
This is not the place to tell about possible bugs in some implementation. Please refer to the issue tracker in the respective repository of your domain.

**Comment by johan-boule** [comment]
OK, thanks, I will suggest on Conan's repository to combine the the 2 license expressions into one with and AND operator:

`GPL-2.0-only WITH Classpath-exception-2.0 AND GPL-2.0-only WITH OpenJDK-assembly-exception-1.0`

Perhaps multiple WITH clauses are allowed, but my guess is that would mean the same.

**Pull request merged**